### PR TITLE
Adding pagination tests

### PR DESCRIPTION
**Pull request by CMurrell148**
(https://dev.azure.com/dfe-gov-uk/Academies-and-Free-Schools-SIP/_workitems/edit/177784)

## Changes

Added tests for pagination mainly of the search page but also things like return to the home screen. 

## Checklist

- [x] Pull request attached to the appropriate user story in Azure DevOps
- [x] Testing complete - all manual and automated tests pass

### CODE REVIEW

File: tests/NewCypress/cypress/e2e/regression/footer.cy.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+        cy.login();
+    });
+
+    it("Should check that the home page footer bar privacy link is present and functional", () => {
+        footerLinks
+            .checkPrivacyLinkPresent()
+            .clickPrivacyLink()
+
+        paginationPage
+            .checkImAtTheCorrectUrl('privacy')
```
Comment (dneed-nimble) [inline tests/NewCypress/cypress/e2e/regression/footer.cy.ts:16]: Could the naming be a bit more verbose? checkImAtTheCorrectUrl, what is Im here, pedeantic I know but CheckCurrentUrlIsCorrect or something akin may give more clarity.

File: tests/NewCypress/cypress/e2e/regression/pagination.cy.ts [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        paginationPage.getTotalPaginationButtons().should('be.greaterThan', 1);
+    });
+
+    it('Should navigate to a specific page when a page number is clicked', () => {
+        const targetPage = 3;
+
+        paginationPage
+            .getResults()
+            .then(initialResults => {
+                const initialFirstResultText = initialResults
```
Comment (dynamictulip) [inline tests/NewCypress/cypress/e2e/regression/pagination.cy.ts:21]: This variable name is a little tough to understand, can we change it to something clearer? Maybe `firstPageFirstResult`?

File: tests/NewCypress/cypress/e2e/regression/pagination.cy.ts [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+                    .text();
+
+                paginationPage
+                    .clickPageNumber(targetPage);
+
+                paginationPage
+                    .getResults()
+                    .first()
+                    .should('not.have.text', initialFirstResultText);
+            });
```
Comment (dynamictulip) [inline tests/NewCypress/cypress/e2e/regression/pagination.cy.ts:32]: This test ensures that we're on a different page to the first result page but not necessarily that we're on the third result page. There's a few different ways to tell which page we're actually on that we could test:
- the url ends in `/search?keywords=west&pagenumber=3`
- the page title is `Search (page 3 of 3)`
- the current page is highlighted in the pagination bar at the bottom (the element has `govuk-pagination__item--current`)

Comment (dynamictulip) [reply tests/NewCypress/cypress/e2e/regression/pagination.cy.ts:32]: We could fold this logic into the pageobject model - `paginationPage.checkCurrentPageIs(3)` which would simplify all the tests and allow us to easily test multiple navigations

e.g.
```
paginationPage.checkCurrentPageIs(1)

paginationPage.clickPageNumber(3)
paginationPage.checkCurrentPageIs(3)

paginationPage.clickPageNumber(2);
paginationPage.checkCurrentPageIs(2)
```

File: tests/NewCypress/cypress/e2e/regression/pagination.cy.ts [hunk 4]
```diff
@@ -0,0 +1,82 @@
+import paginationPage from "../../pages/paginationPage";
+
+describe('Pagination Tests', () => {
+
+
+    beforeEach(() => {
+        cy.login()
+        cy.visit('/search?keywords=west')
```
Comment (dynamictulip) [inline tests/NewCypress/cypress/e2e/regression/pagination.cy.ts:8]: We're only testing one search result which currently has three result pages. It would be good to test a couple of scenarios with different numbers of page results, especially as the ui behaves differently depending on whether there are 1, 2, 3 or more pages of results

File: tests/NewCypress/cypress/e2e/regression/search-results-page.cy.ts [hunk 5]
```diff
@@ -34,7 +34,7 @@ describe("Testing the components of the search results page", () => {
             .checkSearchResultsReturned("East")
     });
 
-    it.only("Validates that it reutrns the amount of results stated in the search text", () => {
+    it("Validates that it reutrns the amount of results stated in the search text", () => {
```
Comment (dynamictulip) [inline tests/NewCypress/cypress/e2e/regression/search-results-page.cy.ts:37]: Missed this one before, small typo
```suggestion
    it("Validates that it returns the amount of results stated in the search text", () => {
```

File: tests/NewCypress/cypress/pages/paginationPage.ts [hunk 6]
```diff
@@ -2,52 +2,67 @@ class PaginationPage {
...
+    public clickPrevious(): this {
+        const getPreviousButtonElement = () => cy.contains('Previous'); 
+        getPreviousButtonElement().click();
         return this;
     }
 
-    public goToPage(pageNumber: string) {
-
-        cy.get(`page-${pageNumber}`).click();
+    public shouldNextButtonBeDisabled(): this {
```
Comment (dynamictulip) [inline tests/NewCypress/cypress/pages/paginationPage.ts:44]: This (and `shouldPreviousButtonBeDisabled`) don't follow the "check" naming convention.

File: tests/NewCypress/cypress/pages/paginationPage.ts
Comment (dynamictulip) [inline tests/NewCypress/cypress/pages/paginationPage.ts:1]: A few functions in this file aren't being used, are they being used in future work?

File: tests/NewCypress/cypress/e2e/regression/pagination.cy.ts (hunk 2 above)
Comment (CMurrell148) [reply tests/NewCypress/cypress/e2e/regression/pagination.cy.ts:21]: Now updated - also changed some similar ones to be a bit clearer.

File: tests/NewCypress/cypress/pages/paginationPage.ts (hunk 6 above)
Comment (CMurrell148) [reply tests/NewCypress/cypress/pages/paginationPage.ts:44]: Now updated this and also the next button function to follow the above convention

File: tests/NewCypress/cypress/e2e/regression/pagination.cy.ts (hunk 3 above)
Comment (CMurrell148) [reply tests/NewCypress/cypress/e2e/regression/pagination.cy.ts:32]: Now implemented the above functionality

File: tests/NewCypress/cypress/e2e/regression/pagination.cy.ts (hunk 4 above)
Comment (CMurrell148) [reply tests/NewCypress/cypress/e2e/regression/pagination.cy.ts:8]: Now added another test that paginates through a search page with a large number of results

File: tests/NewCypress/cypress/e2e/regression/pagination.cy.ts [hunk 7]
```diff
@@ -0,0 +1,133 @@
+import paginationPage from "../../pages/paginationPage";
+
+describe('Pagination Tests', () => {
+
```
Comment (dynamictulip) [inline tests/NewCypress/cypress/e2e/regression/pagination.cy.ts:4]: Missing test cases:

- first page has no previous button
- last page has no next button
- 1 page of results:
  - has no next button and no previous button and only "1" on the bar
- first page, last page and page either side of current page are shown on the bar (e.g. for [page 30 of "trust" search](https://test.find-information-academies-trusts.education.gov.uk/search?keywords=trust&pagenumber=30) you can see `1, ⋯, 29, 30, 31, ⋯ and 73` but for [page 2 of "trust" search](https://test.find-information-academies-trusts.education.gov.uk/search?keywords=trust&pagenumber=2) you can only see `1, 2, 3, ⋯ and 73`)

File: tests/NewCypress/cypress/pages/paginationPage.ts
Comment (dynamictulip) [reply tests/NewCypress/cypress/pages/paginationPage.ts:1]: Unused functions: 
- `getPagination()`
- `checkTheNextPageButtonIsDisabled()`
- `checkThePreviousPageButtonIsDisabled()`
- `iterateThroughPaginationButtons()`

File: tests/NewCypress/cypress/e2e/regression/pagination.cy.ts [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+                .should('not.have.text', firstPageFirstResultText);
+        });
+    });
+
+    it('Checks that the previous page button is not present on the first page of results', () => {
+        paginationPage
+            .checkPreviousButtonNotPresent()
+    });
+
+    it('Checks that the previous page button is not present on the first page of results', () => {
```
Comment (dynamictulip) [inline tests/NewCypress/cypress/e2e/regression/pagination.cy.ts:71]:
```suggestion
    it('Checks that the next page button is not present on the last page of results', () => {
```

File: tests/NewCypress/cypress/e2e/regression/pagination.cy.ts [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+            const secondPageFirstResultText = secondPageResults.first().text();
+
+            paginationPage.clickPrevious();
+
+            paginationPage.getResults().first().should('not.have.text', secondPageFirstResultText);
+        });
+    });
+
+
+    it('Should iterate through all pagination pages and verify results are different', () => {
```
[the remark this answers is not in the document]
Comment (dynamictulip) [reply tests/NewCypress/cypress/e2e/regression/pagination.cy.ts:108]: Earlier test `Should navigate to a specific page when a page number is clicked on a large result page` can have all the `.should('not.have.text', firstPageFirstResultText);`s removed

File: tests/NewCypress/cypress/e2e/regression/pagination.cy.ts (hunk 8 above)
Comment (CMurrell148) [reply tests/NewCypress/cypress/e2e/regression/pagination.cy.ts:71]: Now updated this to say Next instead of previous - My bad I missed this one

File: tests/NewCypress/cypress/e2e/regression/pagination.cy.ts [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+        paginationPage.getResults().then(firstPageFirstResult => {
+            const firstPageFirstResultText = firstPageFirstResult.first().text();
+
+            paginationPage
+                .clickNext();
+
+            paginationPage
+                .getResults()
+                .first()
+                .should('not.have.text', firstPageFirstResultText);
```
[the remark this answers is not in the document]
Comment (CMurrell148) [reply tests/NewCypress/cypress/e2e/regression/pagination.cy.ts:62]: now added functionality to check its at the page it would expect when paginating back and fourth.

File: tests/NewCypress/cypress/e2e/regression/pagination.cy.ts (hunk 7 above)
Comment (CMurrell148) [reply tests/NewCypress/cypress/e2e/regression/pagination.cy.ts:4]: The first two should already be there but I just added a test for the last scenario

File: tests/NewCypress/cypress/e2e/regression/pagination.cy.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (CMurrell148) [reply tests/NewCypress/cypress/e2e/regression/pagination.cy.ts:103]: Implemented the suggested solution

File: tests/NewCypress/cypress/e2e/regression/pagination.cy.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (dynamictulip) [reply tests/NewCypress/cypress/e2e/regression/pagination.cy.ts:41]: Awesome, we con remove this part then
```
paginationPage
  .getResults()
  .first()
  .should('not.have.text', firstPageFirstResultText);
```

[the remark this answers is not in the document]
Comment (dynamictulip) [reply tests/NewCypress/cypress/e2e/regression/pagination.cy.ts:41]: And also the first part -
```
        paginationPage.getResults().then(firstPageFirstResult => {
            const firstPageFirstResultText = firstPageFirstResult.first().text();

```

File: tests/NewCypress/cypress/e2e/regression/pagination.cy.ts [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+
+    it('Checks that I see the pages I would expect mid pagination and dont see the ones that should be hidden', () => {
+        cy.visit('/search?keywords=tru&pagenumber=30')
+
+        paginationPage
+            .checkExpectedPageNumberInPaginationBar(29)
+            .checkExpectedPageNumberInPaginationBar(31)
+            .checkResultIsNotInPaginationBar(4)
+            .checkResultIsNotInPaginationBar(72)
+    });
```
Comment (dynamictulip) [inline tests/NewCypress/cypress/e2e/regression/pagination.cy.ts:76]: Can we add all the boundary conditions to this test?
So 
```
        paginationPage
            .checkExpectedPageNumberInPaginationBar(1)
            .checkResultIsNotInPaginationBar(2)
            .checkResultIsNotInPaginationBar(28)
            .checkExpectedPageNumberInPaginationBar(29)
            .checkExpectedPageNumberInPaginationBar(30)
            .checkExpectedPageNumberInPaginationBar(31)
            .checkResultIsNotInPaginationBar(32)
            .checkResultIsNotInPaginationBar(72)
            .checkExpectedPageNumberInPaginationBar(73)
```

Comment (CMurrell148) [reply tests/NewCypress/cypress/e2e/regression/pagination.cy.ts:81]: Now updated

**Review by dynamictulip — APPROVED** [review]
